lib/LanguageServerRename/Model/Renamer.php

```diff
@@ -0,0 +1,314 @@
+        $this->workspace = $workspace;
+        $this->renameVariable = $renameVariable;
+        $this->nodeUtils = $nodeUtils;
+    }
+
+    public function prepareRename(TextDocumentItem $textDocument, Position $position): ?Range
+    {
+        [$offset, $node] = $this->documentAndPositionToNodeAndOffset($textDocument, $position);
+        
+        if ($this->canRenameNode($node)) {
+            return $this->nodeUtils->getNodeNameRange($node);
+        }
+
+        return null;
+    }
+
+    public function rename(TextDocumentItem $textDocument, Position $position, string $newName): ?WorkspaceEdit
+    {
+        [$offset, $node] = $this->documentAndPositionToNodeAndOffset($textDocument, $position);
+        
+        $phpactorDocument = TextDocumentBuilder::create($textDocument->text)
+            ->uri($textDocument->uri)
+            ->language($textDocument->languageId ?? 'php')
+            ->build();
+
+        if ($this->canRenameNode($node))
+            return $this->renameNode($phpactorDocument, $offset, $node, $this->nodeUtils->getNodeNameText($node, $textDocument->text), $newName);
+        
+        return null;
+    }
+    
+    private function renameNode(TextDocument $phpactorDocument, ByteOffset $offset, Node $node, string $oldName, string $newName): ?WorkspaceEdit
+    {
+        if (empty($oldName)) {
+            return null;
+        }
+        $locations = [];
+        try {
+            $potentialLocation = $this->definitionLocator->locateDefinition($phpactorDocument, $offset);
+            $locations[] = new Location($potentialLocation->uri(), $potentialLocation->offset());
+        } catch (CouldNotLocateDefinition $notFound) {
+            // ignore the missing definition
+        }
+        
+        $start = microtime(true);
+        $count = 0;
+        foreach ($this->finder->findReferences($phpactorDocument, $offset) as $potentialLocation) {
+            if (!$potentialLocation->isSurely()) {
+                continue;
+            }
+            
+            $locations[] = $potentialLocation->location();
+
+            if ($count++ % 100 === 0 && $count > 0) {
+                $this->clientApi->window()->showMessage()->info(sprintf(
+                    '... scanned %s references confirmed %s ...',
+                    $count - 1,
+                    count($locations)
+                ));
+            }
+
+            if (microtime(true) - $start > $this->timeoutSeconds) {
+                $this->clientApi->window()->showMessage()->info(sprintf(
+                    'Reference find stopped, %s/%s references confirmed but took too long (%s/%s seconds).',
+                    count($locations),
+                    $count,
+                    number_format(microtime(true) - $start, 2),
+                    $this->timeoutSeconds
+                ));
+                return $this->locationsToWorkspaceEdit($locations, $oldName, $newName);
+            }
+
+            // I can't see how I can yield here.
```

Comment:
the public `rename` method needs to return a `Promise` and this method needs to be a [co-routine](https://amphp.org/amp/coroutines/). Basically you need to wrap it in `call`:

```
/**
 * @return Promise<WorkspaceEdit|null>
 */
private function doRename($1, $2): Promise
{
    return call(function () use ($1, $2) {
        // ...
        yield delay(0); // maybe nicer to use the function rather than the class.
    });
}
```